PROTOCOL.md

```diff
@@ -0,0 +1,75 @@
+
+The communication is mediated by a network of `nodes`. The network is responsible for ensuring that messages are delivered between clients. Since clients are not necessarily connected to the network at the same time, the network must be able to store messages while they are in transit.
+
+The protocol is layered to satisfy various operational and security requirements. For example, the clients are responsible for all the encryption and authentication, and key management. The nodes generally see only encrypted content with no indication of identities involved.
+
+## Network Layer
+
+Messages are opaque to the node (usually encrypted), i.e. nodes do not read/interpret the contents of the messages. Unencrypted message headers are readable and can be processed by the nodes.
+
+Consequently a message is just an opaque `payload` to the nodes, wrapped in an `envelope` before being passed to the network by the clients. Besides the payload, an envelope also carries a `timestamp` and `topic`.
+
+Envelopes are always associated with a `topic`. Topics are used to group related envelopes. A topic is a required parameter of all interactions with the network layer, e.g. "add this envelope to this topic", "give me all envelopes from this topic with timestamp between X and Y", or "send me all new envelopes for this topic".
+
+Nodes employ the [Waku relay protocol](https://rfc.vac.dev/spec/11/) to propagate envelopes through the network and the [Waku store protocol](https://rfc.vac.dev/spec/13/) to make envelopes persist.
+
+## Client Layer
+
+Clients connect to an arbitrary node through a GRPC-based [message API](https://github.com/xmtp/proto/blob/main/proto/message_api/v1/message_api.proto) and interact with the network as a whole through this node. The message API is the interface separating the client layer from the network layer. Conversations are represented at the network layer as topics.
+
+The client layer uses protobuf for encoding of all its structures. The definitions are maintained in the [xmtp/proto repository](https://github.com/xmtp/proto/blob/main/proto).
+
+### Keys
+
+A client is associated with a blockchain account through a set of public/private keys (EC secp256k1) generated when the account is registered with the XMTP network. First there is the identity key that serves as a proxy for the account key and is signed by the account key to establish its authenticity. Then there are one or more pre-keys that are used for message encryption. The pre-keys are signed by the identity key to prove authenticity. The two-level key structure is there to facilitate more frequent key rotation without requiring frequent account key signing. The key structure and naming are modeled after [Signal's X3DH protocol](https://signal.org/docs/specifications/x3dh/#the-x3dh-protocol).
```

Comment:
What do you think about this change, @mkobetic? =)

```suggestion
A client is associated with a blockchain account through a set of public/private keys (EC secp256k1) generated when the account is registered with the XMTP network. First there is the identity key that serves as a proxy for the account key and is signed by the account key to establish its authenticity. Then there are one or more pre-keys that are used for invitation encryption. The pre-keys are signed by the identity key to prove authenticity. The two-level key structure is there to facilitate more frequent key rotation without requiring frequent account key signing. The key structure and naming are modeled after [Signal's X3DH protocol](https://signal.org/docs/specifications/x3dh/#the-x3dh-protocol).
```

Reply:
Hm, you are right @jhaaaa that could be misleading. But I want to be a bit more general/vague here. So how about I change `message encryption` to `payload encryption and signing`?

Reply:
Thank you for your patient consideration, @mkobetic! I like your suggested change!